pubsub/gcloud/aio/pubsub/subscriber.py

```diff
@@ -0,0 +1,229 @@
+    from gcloud.aio.pubsub.subscriber_client import SubscriberClient
+    from gcloud.aio.pubsub.subscriber_message import SubscriberMessage
+    from gcloud.aio.pubsub.metrics_agent import MetricsAgent
+
+    log = logging.getLogger(__name__)
+
+    if TYPE_CHECKING:
+        MessageQueue = asyncio.Queue[Tuple[SubscriberMessage,  # pylint: disable=unsubscriptable-object
+                                           float]]
+    else:
+        MessageQueue = asyncio.Queue
+    ApplicationHandler = Callable[[SubscriberMessage], Awaitable[None]]
+
+    class AckDeadlineCache:
+        def __init__(self, subscriber_client: SubscriberClient,
+                     subscription: str, cache_timout: int):
+            self.subscriber_client = subscriber_client
+            self.subscription = subscription
+            self.cache_timeout = cache_timout
+            self.ack_deadline: float = float('inf')
+            self.last_refresh: float = 0
+
+        async def get(self) -> float:
+            if self.ack_deadline == float('inf') or self.cache_outdated():
+                await self.refresh()
+            return self.ack_deadline
+
+        async def refresh(self) -> None:
+            try:
+                sub = await self.subscriber_client.get_subscription(
+                    self.subscription)
+                self.ack_deadline = float(sub['ackDeadlineSeconds'])
+            except Exception as e:
+                log.warning(
+                    'Failed to refresh ackDeadlineSeconds value', exc_info=e)
+            self.last_refresh = time.perf_counter()
+
+        def cache_outdated(self) -> bool:
+            if (time.perf_counter() - self.last_refresh) > self.cache_timeout:
+                return True
+            return False
+
+    async def acker(subscription: str,
+                    ack_queue: 'asyncio.Queue[str]',
+                    subscriber_client: 'SubscriberClient',
+                    ack_window: float,
+                    metrics_client: Optional[MetricsAgent]) -> None:
+        ack_ids: List[str] = []
+        while True:
+            time_budget = ack_window
+            if not ack_ids:
+                ack_ids.append(await ack_queue.get())
+                ack_queue.task_done()
+
+            while time_budget > 0:
+                start = time.perf_counter()
+                try:
+                    ack_id = await asyncio.wait_for(
+                        ack_queue.get(), timeout=time_budget)
+                    ack_ids.append(ack_id)
+                    ack_queue.task_done()
+                except asyncio.TimeoutError:
+                    break
+                time_budget -= (time.perf_counter() - start)
+            try:
+                await subscriber_client.acknowledge(subscription,
+                                                    ack_ids=ack_ids)
+            except Exception as e:
+                log.warning(
+                    'Ack request failed, better luck next batch', exc_info=e)
```

Comment:
@shaundialpad we were discussing potentially unbound growth of `ack_ids`. My last take:

> I wonder what would be the best way to safeguard against it? Worst case scenario is, I guess, google's outage specifically on ack endpoint. In this case we need to make sure we won't consume all the memory? So maybe only allow head N messages from last iteration to let through?

Reply:
@TheKevJames We were actually discussing the growth of `ack_ids` list where we store unsent ids. Our acknowledgement train named `acked` is going full speed and stops only to make a request every `0.3` seconds. If for some reason this request fails – not a bid deal, we just keep all the ids in the list and will send them on next batch. The problem I personally see here is long term outage of google's ack endpoint. In this case `ack_ids` list will keep growing until a) it consumes all memory and crash the app b) ack endpoint is available again and we're good c) we exceed POST body length limit and become a one way application 😱

Reply:
Well isn't `ack_ids` initialized as `asyncio.Queue(maxsize=(max_messages * num_workers))`? If we try to insert `maxsize + 1` messages, the consumer attempting to do so will simply wait. Perhaps I'm misunderstanding something?

Reply:
nonono, `ack_ids` is a simple list that is used as a temporary bucket where acker puts ids once they're taken from the queue. This bucket is emptied on successful ack request and is carried to the next iteration if not

Reply:
Ah! Yes, the `continue` on the next line was missing in my mental model, so I thought you were clearing `ack_ids` in each case. Ah, so we do retry acking the same IDs, I see... Hmm, yeah, this is potentially worrying. Perhaps we need a max size on the `ack_ids` list so that we avoid pulling more off the `ack_queue` when it's too full? I imagine that could solve a potential bug as well: I betcha there's a max size on the `acknowledge` payload before it either rejects the request and/or starts being to slow for our timeout settings anyway. Perhaps we should never try to ack more than `n` messages in one go?

Note that relatedly, messages which are repeatedly failing to ack stay in this list forever -- do we need a max retry count?

Reply:
Interesting also is whether the `acknowledge` will acknowledge a subset of the `ack_ids` or will only acknowledge all-or-nothing. If there is one ID in `ack_ids` that makes the request fail, will it continue to fail on each request until you run out of memory? How do you know which ID is "bad" and needs to be dropped?

Reply:
I like the idea to avoid trying to ack more than `n` messages in one go in that at some point you need to stop requesting more items from the `ack_queue` and "block" on the current batch to cause backpressure in the consumer and producer tasks.

Reply:
@TheKevJames @jonathan-johnston 
acknowledging endpoint is "all-or-nothing" in a sense that it's always "all" if you could get through network, authentication and validation:
>Acknowledging a message whose ack deadline has expired may succeed, but such a message may be redelivered later. Acknowledging a message more than once will not result in an error.

So basically if we got a response there is nothing left to do about those ids. That's why "messages failing to ack" is not even a concept here.

Google does not mention any specific limit for this endpoint. For publishing messages it is 1000 messages or 10MB, so I think it's safe to only keep 1000 messages between post attempts?

@jonathan-johnston acker backpressure is actually an interesting topic. I'd argue we don't need it _mostly_, because if a batch failed to ack, best we can do is to try again. But the priority should always go to the most recent messages, because those have higher chances to succeed in time. If we make acker a "fair" worker that wants to reliably ack everything we effectively will be increasing damage caused by random hiccups.

Reply:
Alright so the limit is ~2744 messages, I've added a simple slice to cover that. Let me know what you think

Reply:
@amartsinovich-dialpad Thanks for digging on that, it's good to know that expired messages are just successful. So the only failure cases we'd see here are rate limiting, network timeouts, and server errors. This makes a lot more sense now and seems safe with the message limit.

Reply:
> However this doesn't seem to solve the case of unbounded list growth, is that right?

I might have pushed the updated version a bit after you checked the code. Check reassignment at line 83, it should solve the unbound list. Maximum one-time list size should be 2500 messages from last iteration + messages we gathered during one batch window.

Reply:
Ah, gotcha, thanks!

Reply:
I like it. I think :P So with no back pressure, worse case scenario is messages will just be organically redelivered. 

Should we add a metric for failed ack requests?

Reply:
I don't see why not